flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -57,11 +79,12 @@ class StopWithSavepoint extends StateWithExecutionGraph {
 
     private final CheckpointScheduling checkpointScheduling;
 
-    private boolean hasFullyFinished = false;
-
-    @Nullable private String savepoint = null;
-
     @Nullable private Throwable operationFailureCause;
+    private boolean hasPendingStateTransition = false;
+
+    // be careful when applying operations on this future that can trigger state transitions,
+    // as several other methods do the same and we mustn't trigger multiple transitions!
+    private final CompletableFuture<String> internalSavepointFuture = new CompletableFuture<>();
```

Comment:
May we add some comment to the `operationsFuture` as well to distinguish the two from each other? The purpose of the `operationFuture` is, as far as I understand, to have a future that completes dependening on the internal state as soon as the subsequent state transition is happening.